### ISSUE TITLE
about the iou of unsupervised co-segmentation   

### ISSUE THREAD

**Issue by zero-zm** [czq142857/BAE-NET#3]
hello, I run the  unsupervised co-segmentation  code  recently. but i just get a very small iou value( such as 37 , 26.8). The same configuration gives me an IOU value of 82 on PointNet. so i want to know did you have any method to improve it? Thanks

**Comment by czq142857** [comment]
If you are using *unsupervised* co-segmentation code, you need to manually rearrange the ground-truth parts to match the corresponding branches, because in unsupervised settings we do not know which part will appear in which branch.

For example, if your network has 4 branches and they output chair seat, back, leg, arm respectively, and your ground-truth parts have another order: leg, seat, back, arm, then you will get a low IOU because the network output and the ground truth do not match.

In summary, please visualize the network segmentations, and assign the correct ground-truth part to the corresponding branch (or branches).

Best,
Zhiqin

**Comment by zero-zm** [comment]
> If you are using _unsupervised_ co-segmentation code, you need to manually rearrange the ground-truth parts to match the corresponding branches, because in unsupervised settings we do not know which part will appear in which branch.
> 
> For example, if your network has 4 branches and they output chair seat, back, leg, arm respectively, and your ground-truth parts have another order: leg, seat, back, arm, then you will get a low IOU because the network output and the ground truth do not match.
> 
> In summary, please visualize the network segmentations, and assign the correct ground-truth part to the corresponding branch (or branches).
> 
> Best,
> Zhiqin

**Comment by zero-zm** [comment]
Many thanks for your reply，I accidentally closed the issue.  But I have other problems: 
first, Can I visualize it through the following code that you provide?
        python test_vox.py
then,  Can you simply demonstrate how to assign the correct ground-truth part to the corresponding branch?

Looking forward to your reply.  
Best wishes

**Comment by czq142857** [comment]
Please read the "Evaluation and Visualization" section of the README.md file.
After visualizing a few segmentations, you should know which branch corresponds to which part. Then you can post-process the raw network outputs by rearranging or grouping the branches.

Best,
Zhiqin

**Comment by zero-zm** [comment]
I've tried it.  In the same environment, the result of 'checkpoint_1shot_1_3000 ' can be well visualized. But when i visualized a result of 'checkpoint_unsup', it reports an error '  AttributeError: 'IMSEG' object has no attribute 'ref_b_point_num'    '

**Comment by czq142857** [comment]
You need to use the correct parameters. For unsupervised experiments you should not use the flag "--supervised" in your terminal commands.

Best,
Zhiqin

**Comment by zero-zm** [comment]
I removed  '--supervised '.   I used the command 'python main.py --pointcloud --L1reg --iteration 200000 --dataset 03001627_vox --data_dir ./data/03001627_chair/ --checkpoint_dir checkpoint_unsup --supervision_list 03001627_test_vox.txt --sample_dir samples/03001627_chair  --real_size 32 --points_per_shape 8192', which reports an error ' AttributeError: 'IMSEG' object has no attribute 'ref_b_point_num' '

**Comment by zero-zm** [comment]
I'm so excited. I have  visualized some segmentation results. but i don't know which part of the file i should modify to post-process the raw network outputs. what exactly do i need to do? Thanks!

**Comment by czq142857** [comment]
Here is a pipeline of what you need to do:
![image](https://user-images.githubusercontent.com/24209790/87358432-ad90f000-c51a-11ea-97e6-51931204194e.png)

For example, you can implement grouping after line 458 in model.py:
pred_part_labels = np.argmax(model_out, axis=1).astype(np.int32)  --line 458
#do the grouping
pred_part_labels = your_grouping_function(pred_part_labels)
...
